substrate/frame/revive/src/storage/meter.rs

```diff
@@ -403,19 +353,32 @@ where
 			Origin::Root => return Ok(Deposit::Charge(Zero::zero())),
 			Origin::Signed(o) => o,
 		};
+
+		// Coalesce charges of the same contract
 		self.charges.sort_by(|a, b| a.contract.cmp(&b.contract));
-		// Coalesce charges of the same contract.
 		self.charges = {
 			let mut coalesced: Vec<Charge<T>> = Vec::with_capacity(self.charges.len());
-			for ch in self.charges {
+			for mut ch in self.charges {
 				if let Some(last) = coalesced.last_mut() {
 					if last.contract == ch.contract {
-						// merge amounts (uses Deposit::saturating_add)
-						last.amount = last.amount.saturating_add(&ch.amount);
-						// prefer terminated state if any entry is terminated (keep whichever is
-						// terminated)
-						if matches!(ch.state, ContractState::Terminated { .. }) {
-							last.state = ch.state.clone();
+						match (&mut last.state, &mut ch.state) {
+							(
+								ContractState::Alive { amount: last_amount },
+								ContractState::Alive { amount: ch_amount },
+							) => {
+								*last_amount = last_amount.saturating_add(ch_amount);
+							},
+							(ContractState::Alive { amount }, ContractState::Terminated) |
+							(ContractState::Terminated, ContractState::Alive { amount }) => {
```

Comment:
this second one can never happen right and is just here to be exhaustive in the match statement?
you can't have a charge following a Terninated ?